File: ROID-Forum-Web-Client/Convert Client Code/GMGWebsite/web/Website/Networking/MessageWriter.dart
library MessageWriter;
import 'dart:typed_data';

class MessageWriter {
  int byteLength;
  List<MessageData> data;
  MessageWriter() {
    data = new List();
    byteLength = 4;
  }
  /*
  ByteData generateByteData() {
    ByteData byteData = new ByteData(byteLength);
    int loc = 0;
    byteData.setUint32(loc, byteLength);
    loc += 4;
    for (int i = 0; i < data.length; i++) {
      MessageData m = data[i];
      m.addToByteData(byteData, loc);
      loc += m.getLength();
    }
    return byteData;
  }*/
  void addToByteData(ByteData byteData, int loc) {
    byteData.setUint32(loc, byteLength);
    loc += 4;
    for (int i = 0; i < data.length; i++) {
      MessageData m = data[i];
      m.addToByteData(byteData, loc);
      loc += m.getLength();
    }
  }
  void addUint8(int value) {
    MessageDataUint8 m = new MessageDataUint8(value);
    data.add(m);
    byteLength += m.getLength();
  }
  void addUint16(int value) {
    MessageDataUint16 m = new MessageDataUint16(value);
    data.add(m);
    byteLength += m.getLength();
  }
  void addInt16(int value) {
      MessageDataInt16 m = new MessageDataInt16(value);
      data.add(m);
      byteLength += m.getLength();
    }
  void addUint32(int value) {
    MessageDataUint32 m = new MessageDataUint32(value);
    data.add(m);
    byteLength += m.getLength();
  }
  void addInt32(int value) {
    MessageDataInt32 m = new MessageDataInt32(value);
    data.add(m);
    byteLength += m.getLength();
  }
  void addFloat64(double value) {
    MessageDataFloat64 m = new MessageDataFloat64(value);
    data.add(m);
    byteLength += m.getLength();
  }
  void addFloat32(double value) {
    MessageDataFloat32 m = new MessageDataFloat32(value);
    data.add(m);
    byteLength += m.getLength();
  }
  void addString(String value) {
    MessageDataString m = new MessageDataString(value);
    data.add(m);
    byteLength += m.getLength();
  }
  /*
  void addBinary(MessageWriter value) {
    MessageDataBinary m = new MessageDataBinary(value);
    data.add(m);
    byteLength += m.getLength();
  }*/
  int getLength() {
    return byteLength;
  }
}






abstract class MessageData {
  void addToByteData(ByteData byteData, int loc);
  int getLength();
}
class MessageDataUint8 implements MessageData {
  int value;
  MessageDataUint8(this.value);
  void addToByteData(ByteData byteData, int loc) {
    byteData.setUint8(loc, value);
  }
  int getLength() {
    return 1;
  }
}
class MessageDataUint16 implements MessageData {
  int value;
  MessageDataUint16(this.value);
  void addToByteData(ByteData byteData, int loc) {
    byteData.setUint16(loc, value);
  }
  int getLength() {
    return 2;
  }
}
class MessageDataInt16 implements MessageData {
  int value;
  MessageDataInt16(this.value);
  void addToByteData(ByteData byteData, int loc) {
    byteData.setInt16(loc, value);
  }
  int getLength() {
    return 2;
  }
}
class MessageDataUint32 implements MessageData {
  int value;
  MessageDataUint32(this.value);
  void addToByteData(ByteData byteData, int loc) {
    byteData.setUint32(loc, value);
  }
  int getLength() {
    return 4;
  }
}
class MessageDataFloat64 implements MessageData {
  double value;
  MessageDataFloat64(this.value);
  void addToByteData(ByteData byteData, int loc) {
    byteData.setFloat64(loc, value);
  }
  int getLength() {
    return 8;
  }
}
class MessageDataFloat32 implements MessageData {
  double value;
  MessageDataFloat32(this.value);
  void addToByteData(ByteData byteData, int loc) {
    byteData.setFloat32(loc, value);
  }
  int getLength() {
    return 4;
  }
}
class MessageDataInt32 implements MessageData {
  int value;
  MessageDataInt32(this.value);
  void addToByteData(ByteData byteData, int loc) {
    byteData.setInt32(loc, value);
  }
  int getLength() {
    return 4;
  }
}
class MessageDataString implements MessageData {
  String value;
  List<int> stringCode;
  MessageDataString(this.value) {
    stringCode = value.codeUnits;
  }
  void addToByteData(ByteData byteData, int loc) {
    byteData.setUint16(loc, value.length);
    for (int i = 0; i < stringCode.length; i++) {
      byteData.setUint16(loc + 2 + i * 2, stringCode[i]);
    }
  }
  int getLength() {
    return 2 + stringCode.length * 2;
  }
}
/*
class MessageDataBinary implements MessageData {
  MessageWriter value;
  MessageDataBinary(this.value) {}
  void addToByteData(ByteData byteData, int loc) {
    value.addToByteData(byteData, loc);
  }
  int getLength() {
    return value.getLength();
  }
}*/

File: ROID-Forum-Server/ROID-Forum-Server/Networking/Message/MessageWriter.cs
using System;
using System.Collections.Generic;
using Soly.Utilities.ByteArray;

namespace ROIDForumServer
{
	public class MessageWriter
	{
		private readonly List<MessageData> _dataArray = new List<MessageData>();
		private uint _byteLength = 0;

		public byte[] ToBuffer()
		{
			ByteArray byteData = new ByteArray((int)_byteLength);
			uint loc = 0;
			foreach (MessageData data in _dataArray)
			{
				data.AddToByteData(byteData, (int)loc);
				loc += data.GetLength();
			}
			return byteData.Buffer;
		}

		public void AddUint8(byte value)
		{
			var data = new MessageDataUint8(value);
			_dataArray.Add(data);
			_byteLength += data.GetLength();
		}

		public void AddInt8(sbyte value)
		{
			var data = new MessageDataInt8(value);
			_dataArray.Add(data);
			_byteLength += data.GetLength();
		}

		public void AddUint16(ushort value)
		{
			var data = new MessageDataUint16(value);
			_dataArray.Add(data);
			_byteLength += data.GetLength();
		}

		public void AddInt16(short value)
		{
			var data = new MessageDataInt16(value);
			_dataArray.Add(data);
			_byteLength += data.GetLength();
		}

		public void AddUint32(uint value)
		{
			var data = new MessageDataUint32(value);
			_dataArray.Add(data);
			_byteLength += data.GetLength();
		}

		public void AddInt32(int value)
		{
			var data = new MessageDataInt32(value);
			_dataArray.Add(data);
			_byteLength += data.GetLength();
		}

		public void AddFloat64(double value)
		{
			var data = new MessageDataFloat64(value);
			_dataArray.Add(data);
			_byteLength += data.GetLength();
		}

		public void AddFloat32(float value)
		{
			var data = new MessageDataFloat32(value);
			_dataArray.Add(data);
			_byteLength += data.GetLength();
		}

		public void AddString(string value)
		{
			var data = new MessageDataString(value);
			_dataArray.Add(data);
			_byteLength += data.GetLength();
		}
        
		public void AddBinary(byte[] value)
		{
			var data = new MessageDataBinary(value);
			_dataArray.Add(data);
			_byteLength += data.GetLength();
		}

		public uint GetLength()
		{
			return _byteLength;
		}
	}
}